drivers/dma/dma_cavs_gpdma.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT intel_cavs_gpdma
+
+#define GPDMA_CTL_OFFSET 0x004
+#define GPDMA_CTL_FDCGB BIT(0)
+
+#include "dma_dw_common.h"
+
+#define LOG_LEVEL CONFIG_DMA_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(dma_cavs_gpdma);
+
+
+/* Device run time data */
+struct cavs_gpdma_data {
+	struct dw_dma_dev_data dw_data;
+};
+
+/* Device constant configuration parameters */
+struct cavs_gpdma_cfg {
+	struct dw_dma_dev_cfg dw_cfg;
+	uint32_t shim;
+};
+
+/* Disables automatic clock gating (force disable clock gate) */
+static void cavs_gpdma_clock_enable(const struct device *dev)
+{
+	const struct cavs_gpdma_cfg *const dev_cfg = dev->config;
+	uint32_t reg = dev_cfg->shim + GPDMA_CTL_OFFSET;
+
+	sys_write32(GPDMA_CTL_FDCGB, reg);
+}
+
+
+int cavs_gpdma_init(const struct device *dev)
+{
+	const struct cavs_gpdma_cfg *const dev_cfg = dev->config;
+
+	/* Disable dynamic clock gating appropriately before initializing */
+	cavs_gpdma_clock_enable(dev);
+
+	/* Disable all channels and Channel interrupts */
+	dw_dma_setup(dev);
+
+	/* Configure interrupts */
+	dev_cfg->dw_cfg.irq_config();
+
+	LOG_INF("Device %s initialized", dev->name);
+
+	return 0;
+}
+
+
+static const struct dma_driver_api cavs_gpdma_driver_api = {
+	.config = dw_dma_config,
+	.reload = dw_dma_reload,
+	.start = dw_dma_transfer_start,
+	.stop = dw_dma_transfer_stop,
+};
+
+#define CAVS_GPDMA_INIT(inst)						\
+									\
+	static struct dw_drv_plat_data dmac##inst = {			\
```

Comment:
Seems like you could get the number of channels out of devicetree and then just declare this as an array.  I know various instantiations are intended to have different numbers of channels.

Also as the initialization values are identical in all cases, this seems like a waste of bytes in .data vs. just putting it in .bss and initializing at runtime by setting class to 6.

Reply:
They aren't always the same in SOF, and this is something I need to port over. They vary slightly per platform